test.js

```diff
@@ -5,57 +5,107 @@ var assert = require('assert');
 
 describe('glob-parent', function() {
   it('should strip glob magic to return parent path', function() {
-    assert.equal(gp('path/to/*.js'), 'path/to');
-    assert.equal(gp('/root/path/to/*.js'), '/root/path/to');
+    assert.equal(gp('.'), '.');
+    assert.equal(gp('.*'), '.');
+    assert.equal(gp('/.*'), '/');
+    assert.equal(gp('/.*/'), '/');
+    assert.equal(gp('a/.*/b'), 'a');
+    assert.equal(gp('a*/.*/b'), '.');
+    assert.equal(gp('*/a/b/c'), '.');
+    assert.equal(gp('*'), '.');
+    assert.equal(gp('*/'), '.');
+    assert.equal(gp('*/*'), '.');
+    assert.equal(gp('*/*/'), '.');
+    assert.equal(gp('**'), '.');
+    assert.equal(gp('**/'), '.');
+    assert.equal(gp('**/*'), '.');
+    assert.equal(gp('**/*/'), '.');
     assert.equal(gp('/*.js'), '/');
     assert.equal(gp('*.js'), '.');
     assert.equal(gp('**/*.js'), '.');
+    assert.equal(gp('{a,b}'), '.');
+    assert.equal(gp('/{a,b}'), '/');
+    assert.equal(gp('/{a,b}/'), '/');
+    assert.equal(gp('(a|b)'), '.');
+    assert.equal(gp('/(a|b)'), '/');
+    assert.equal(gp('./(a|b)'), '.');
+    assert.equal(gp('a/(b c)'), 'a', 'not an extglob');
+    assert.equal(gp('a/(b c)/d'), 'a/(b c)', 'not an extglob');
+    assert.equal(gp('path/to/*.js'), 'path/to');
+    assert.equal(gp('/root/path/to/*.js'), '/root/path/to');
+    assert.equal(gp('chapter/foo [bar]/'), 'chapter');
     assert.equal(gp('path/[a-z]'), 'path');
     assert.equal(gp('path/{to,from}'), 'path');
     assert.equal(gp('path/(to|from)'), 'path');
+    assert.equal(gp('path/(foo bar)/subdir/foo.*'), 'path/(foo bar)/subdir');
     assert.equal(gp('path/!(to|from)'), 'path');
     assert.equal(gp('path/?(to|from)'), 'path');
     assert.equal(gp('path/+(to|from)'), 'path');
     assert.equal(gp('path/*(to|from)'), 'path');
     assert.equal(gp('path/@(to|from)'), 'path');
+    assert.equal(gp('path/!/foo'), 'path/!');
+    assert.equal(gp('path/?/foo'), 'path', 'qmarks must be escaped');
+    assert.equal(gp('path/+/foo'), 'path/+');
+    assert.equal(gp('path/*/foo'), 'path');
+    assert.equal(gp('path/@/foo'), 'path/@');
+    assert.equal(gp('path/!/foo/'), 'path/!/foo');
+    assert.equal(gp('path/?/foo/'), 'path', 'qmarks must be escaped');
+    assert.equal(gp('path/+/foo/'), 'path/+/foo');
+    assert.equal(gp('path/*/foo/'), 'path');
+    assert.equal(gp('path/@/foo/'), 'path/@/foo');
     assert.equal(gp('path/**/*'), 'path');
     assert.equal(gp('path/**/subdir/foo.*'), 'path');
+    assert.equal(gp('path/subdir/**/foo.js'), 'path/subdir');
+    assert.equal(gp('path/!subdir/foo.js'), 'path/!subdir');
+  });
+
+  it('should respect escaped characters', function() {
+    assert.equal(gp('path/\\*\\*/subdir/foo.*'), 'path/\\*\\*/subdir');
+    assert.equal(gp('path/\\[\\*\\]/subdir/foo.*'), 'path/\\[\\*\\]/subdir');
+    assert.equal(gp('path/\\*(a|b)/subdir/foo.*'), 'path');
```

Comment:
this one can be disregarded now, but I'm interested in perspectives on the other questions.